modules/deployment-scripts/create-agw-kv-certificate/main.bicep

```diff
@@ -0,0 +1,229 @@
+    }
+  }
+  kind: 'AzureCLI'
+  dependsOn: [
+    rbacKv
+    rbacAppGw
+  ]
+  properties: {
+    forceUpdateTag: forceUpdateTag
+    azCliVersion: azCliVersion
+    timeout: timeout
+    retentionInterval: retention
+    environmentVariables: [
+      {
+        name: 'RG'
+        value: resourceGroup().name
+      }
+      {
+        name: 'akvName'
+        value: akvName
+      }
+      {
+        name: 'agwName'
+        value: agwName
+      }
+      {
+        name: 'agwCertType'
+        value: agwCertType
+      }
+      {
+        name: 'certName'
+        value: certName
+      }
+      {
+        name: 'initialDelay'
+        value: initialScriptDelay
+      }
+      {
+        name: 'certificateWait'
+        value: agwCertImportDelay
+      }
+    ]
+    scriptContent: '''
+      #!/bin/bash
+      set -e
+
+      echo "Waiting on Identity RBAC replication ($initialDelay)"
+      sleep $initialDelay
+
+      echo "Creating AKV Cert $certName"
+      az keyvault certificate create --vault-name $akvName -n $certName -p "$(az keyvault certificate get-default-policy | sed -e s/CN=CLIGetDefaultPolicy/CN=${certName}/g )";
+
+      echo "Waiting for AKV Cert Generation ($certificateWait)"
+      sleep $certificateWait
+
+      echo "getting akv secretid for $certName to add to $agwName";
+      versionedSecretId=$(az keyvault certificate show -n $certName --vault-name $akvName --query "sid" -o tsv);
+      unversionedSecretId=$(echo $versionedSecretId | cut -d'/' -f-5) # remove the version from the url;
+
+      case $agwCertType in
+        none)
+          echo "AppGw Certificate Creation not configured"
+          ;;
+
+        root-cert)
+          echo "Creating root certificate in application gateway";
+          rootcertcmd="az network application-gateway root-cert create --gateway-name $agwName -g $RG -n $certName --key-vault-secret-id $unversionedSecretId";
+          echo $rootcertcmd #Inspecting az command is a great DEBUG for the script output
+          $rootcertcmd
+          ;;
+
+        ssl-cert)
+          echo "Creating ssl certificate in application gateway";
+          fecertcmd="az network application-gateway ssl-cert create --gateway-name $agwName -g $RG -n $certName --key-vault-secret-id $unversionedSecretId";
+          echo $fecertcmd #Inspecting az command is a great DEBUG for the script output
+          $fecertcmd
+          ;;
+
+      esac
+
+      jsonOutputString=$(jq -n --arg cn "$certName" --arg sid "$unversionedSecretId" --arg vsid "$versionedSecretId" --arg ct "$agwCertType" '{certName: $cn, certSecretId: {versioned: $vsid, unversioned: $sid }, agwCertType: $ct}')
+      echo $jsonOutputString > $AZ_SCRIPTS_OUTPUT_PATH
+    '''
+    cleanupPreference: cleanupPreference
+  }
+}]
+
+@description('Array of info from each Certificate Deployment Script')
+output createdCertificates array = [for (certName, i) in certNames: {
+  certName: certName
+  DeploymentScriptName: createImportCert[i].name
```

Comment:
I would limit outputs to those that are relevant to the scenario... if debugging is needed then a few thoughts
1) the script isn't robust enough - e.g. you wouldn't need this info output for a deployment of another resource type...
2) the subscription will still have details of the script that can be used to help debug

IOW, we're letting the implementation (of the deploymentScript) bleed through the module.

Reply:
I'll remove the datetime info, but the KV id's will be useful to modules that consume this.

Reply:
agreed - was thinking more of just the deploymentScript goo...